README.md
#### Package moved permanently

As of 2022-03-03, the package has been moved to internal repository. All further development of this library, including new features and bug fixes, will continue in the new repository.

# COAM thin client

A thin client to perform certain requests in COAM. This is **not** an exhaustive implementation
of COAM API. 

If you like to add a method (or tweak something), please open a pull-request. We love pull-requests!
But we also love solving issues, so in case you don't really have the time for a pull-request right now,
maybe you can [just tell us what you'd like](https://github.com/Cimpress/coam-client/issues)? 

Or maybe you just feel like [writing us a mail](mailto:TrdelnikSquad@cimpress.io)? Yes, you guessed right, we also love reading mails ;) 


#### CoamClient class
A simple client to make API calls to COAM. It has a few additional features that can be controlled
by the `options` parameter passed in the constructor.

    const client = new CoamClient(options);
    
###### Accepted options are:
* **accessToken** (required) - The access token to use to authenticate the API calls
* _baseUrl_ - Base COAM API URL. Defaults to https://api.cimpress.io
* _retryAttempts_ - How many tries to retry the requests in case of failure*. Defaults to 2.
* _retryDelayInMs_ - How many milliseconds to wait between retires. Defaults to 200.
* _retryOnForbidden_ - A flag specifying if a retry has to be performed in case of 'Forbidden' response. Defaults to 'true'
* _debugFunction_ - A function to call to provide debug information. Defaults to none.
* _errorFunction_ - A function to call on error. Defaults to console.err;
* _timeout_ - A value setting the timeout of the underlying HTTP calls to COAM. Defaults to 8000 (8 seconds). Set it to 0 for no timeout.
* _skipCache_ - A flag specifying if cache should be skipped. Defaults to 'false'

###### Provided methods 
* buildGroupUrlFromId(groupId)
* hasPermission(principal, resourceType, resourceIdentifier, permission)
* grantRoleToPrincipal(groupUrl, principal, roleName)
* getGroupInfo(groupUrl)
* setAdminFlag(groupId, principal, isAdmin)
* removeUserRole(groupId, principal, role)
* addUserRole(groupId, principal, role)
* modifyUserRoles(groupId, principal, rolesChanges)
* addGroupMember(groupId, principal, isAdmin)
* removeGroupMember(groupId, principal)
* getRoles()
* findPrincipals(query)
* createGroup(name, description)
* getPrincipal(accessToken, principal)
* removeGroup(groupId)
* findGroups(resourceType, resourceIdentifier)
* removeResourceFromGroup(groupId, resourceType, resourceId)
* addResourceToGroup(groupId, resourceType, resourceId)
* getUserPermissionsForResourceType(principal, resourceType, include, permissionFilters)
* getUsersWithPermission(resourceType, resourceIdentifier, permission)
* getUsersWithResource(resourceType, permissionFilters)
* createGroupWithUser(principalToCreateGroup, principalToAddToGroup, groupName, groupDescription, rolesToAdd, resourcesToAdd)

#### Direct helper functions
In some cases, it is easier to simply call a function to perform the required action
without the need for creating a specific client. The following helper methods satisfy these
needs by creating a default client and performing the necessary requests. 

Here is a list of helpers (name + required parameters):

* **buildGroupUrlFromId**(groupId)
* **hasPermission**(accessToken, principal, resourceType, resourceIdentifier, permission) 
* **grantRoleToPrincipal** (accessToken, groupUrl, principal, roleName) 
* **getGroupInfo** (accessToken, groupUrl) 
* **setAdminFlag** (accessToken, groupId, principal, isAdmin) 
* **removeUserRole** (accessToken, groupId, principal, role) 
* **addUserRole** (accessToken, groupId, principal, role) 
* **modifyUserRoles** (accessToken, groupId, principal, rolesChanges) 
* **addGroupMember** (accessToken, groupId, principal, isAdmin) 
* **removeGroupMember** (accessToken, groupId, principal) 
* **getRoles** (accessToken) 
* **findPrincipals** (accessToken, query) 
* **getPrincipal** (accessToken, principal)
* **createGroup** (accessToken, name, description) 
* **removeGroup** (accessToken, groupId) 
* **findGroups** (accessToken, resourceType, resourceIdentifier) 
* **removeResourceFromGroup** (accessToken, groupId, resourceType, resourceId) 
* **addResourceToGroup** (accessToken, groupId, resourceType, resourceId) 
* **getUserPermissionsForResourceType** (accessToken, principal, resourceType, include, permissionFilters)
* **getUsersWithPermission** (accessToken, resourceType, resourceIdentifier, permission)
* **getUsersWithResource** (accessToken, resourceType, permissionFilters)
* **createGroupWithUser** (accessToken, principalToCreateGroup, principalToAddToGroup, groupName, groupDescription, rolesToAdd, resourcesToAdd)

**Important**: All helper methods are using the CoamClient class with default settings (see above).  


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.5.3] - 2022.03.03
### Changed
- The package is undeprecated and moved to internal repository.

## [0.5.2] - 2022.03.02
### Changed
- The package is deprecated. Please view README.md for more information.

## [0.5.1] - 2022.02.17
### Fixed
- Updated axios to 0.26 and axios-retry to 3.2.4

## [0.4.2] - 2022.01.03
### Fixed
- Fixed error in createGroupWithUser related to checking status of error thrown

## [0.4.0] - 2021.12.09
### Fixed
- Updated all dependencies

## [0.3.2] - 2021.12.07
### Fixed
- Fixed incorrect TypeScript typings for createGroupWithUser

## [0.3.1] - 2021.12.07
### Fixed
- Fixed incorrect TypeScript typings for createGroupWithUser

## [0.3.0] - 2021.12.06
### Added
- TypeScript typings are now bundled with the package

## [0.2.8] - 2020.09.09
### Added
- Add function `getUserPermissionsForResourceTypes` to allow getting permissions for different resource types at once.

## [0.2.5] - 2020.06.18
### Added
- Add function `getUsersWithResource` to allow querying the COAM search endpoint for users with a list of permissions to a resource

## [0.2.4] - 2019.12.06
### Added
- Typescript typings

## [0.1.16] - 2019.09.10
### Changed
- Updated dependencies to address security vulnerabilities

## [0.1.15] - 2019.07.24
### Changed
- Wrap axios errors to avoid logging the http client's buffer.

## [0.1.14] - 2019.06.18
### Changed
- Allow getUsersWithPermission to be called with an undefined resourceIdentifier, as this parameter is now optional in the COAM API

## [0.1.13] - 2019.06.06
### Changed
- Downgrade axios until a fix for axios-retry is avaialble

## [0.1.12] - 2019.06.06
### Changed
- Updated deps to address security concerns

## [0.1.10] - 2019.03.25
### Added
- Include a convenience method for constructing a COAM group URL from an ID (buildGroupUrlFromId())

## [0.1.7] - 2019.02.06
### Added
- Fixed incorrect response of createGroup method

## [0.1.6] - 2019.01.17
### Added
- Pass shouldResetTimeout=true to axiosRetry to ensure better retry results
- Fixed searching for users with a specified permission for a resource.

## [0.1.4] - 2019.01.17
### Added
- Allow to search for users with a specified permission for a resource.

## [0.1.3] - 2018.12.05
### Fixed
- Fixed babel runtime issues, upgraded to babel 7

## [0.1.2] - 2018.12.05
### Added
- Get unique principal helper

## [0.1.0] - 2018.12.05
### Added
- Initial release

## [0.0.x] - 2018.11.30
### Added
- Initial version (PROTOTYPE)


tests/todo.tests.js

describe('TO', function() {
    it('DO', function() {
        console.log('!');
    });
});


tests/index.tests.js
const indexExports = require('../src/index');
const expect = require('chai').expect;

describe('Index', function() {
    it('exports are as expected', function() {
        const r = Object.keys(indexExports);
	r.sort();
	
	const e = ['CoamClient',
            'getRoles',
            'grantRoleToPrincipal',
            'addUserRole',
            'removeUserRole',
            'modifyUserRoles',
            'setAdminFlag',
            'createGroup',
            'removeGroup',
            'getGroupInfo',
            'addGroupMember',
            'removeGroupMember',
            'group56',
            'buildGroupUrlFromId',
            'addResourceToGroup',
            'removeResourceFromGroup',
            'hasPermission',
            'getUserPermissionsForResourceType',
            'findGroups',
            'findPrincipals',
            'getPrincipal'];
	e.sort();

        expect(r).to.deep.equal(e);
    });
});


index.d.ts
type KeyValue<T> = {[key: string]: T};

type Profile = any;

export interface CoamClientOptions {
  baseUrl?: string;
  timeout?: number;
  accessToken?: string;
  retryAttempts?: number;
  retryDelayInMs?: number;
  retryOnForbidden?: boolean;
  debugFunction?: (message?: any, ...optionalParams: any[]) => void;
  errorFunction?: (message?: any, ...optionalParams: any[]) => void;
}

interface GroupPrimitiveArray {
  groups: GroupPrimitive[];
}

interface GroupPrimitive {
  name: string;
  description: string;
  id: number;
  created_by?: string;
  created_at?: string;
}

interface Group extends GroupPrimitive {
  members: GroupMember[];
  resources: Resource[];
  permissions: KeyValue<ResourceIdentifierPermissions[]>;
}

interface GroupMember {
  principal: string;
  is_admin: boolean;
  roles: string[];
  description?: string;
}

interface ResourceArray {
  resources: Resource[];
}

interface Resource {
  resource_type: string;
  resource_identifier: string;
}

interface ResourceIdentifierPermissions {
  identifier: string;
  permissions: string[];
  resource_type: string;
}

interface RoleArray {
  roles: Role[]
}

interface Role {
  name: string;
  description: string;
  permissions: ResourceTypePermissionDescription[];
  id: string;
  created_at?: string;
}

interface ResourceTypePermissionDescription {
  resource_type: string;
  name: string;
  description: string;
}

interface CanonicalPrincipalPrimitiveArrayObject {
  canonical_principals: CanonicalPrincipalPrimitive[];
}

interface CanonicalPrincipalPrimitive {
  is_client?: boolean;
  account_id?: string;
  canonical_principal: string;
  is_pending: boolean;
  profiles: Profile[];
}

interface Principal {
  groups: PrincipalGroup[];
  permissions: KeyValue<ResourceIdentifierPermissions[]>;
  canonical_principal: string;
  is_client: boolean;
  is_pending: boolean;
  profile?: Profile;
  account_id?: string;
}

interface PrincipalGroup {
  name: string;
  description: string;
  id: string;
  created_by?: string;
  created_at?: string;
  is_admin: boolean;
}

interface ByResourceSearchResult {
  canonical_principals: CanonicalPrincipalSearchResult[];
}

interface CanonicalPrincipalSearchResult {
  canonical_principal: string;
  resource_type: string;
  is_client: boolean;
  permissions: KeyValue<string[]>;
}

export declare class CoamClient {
  constructor(options: CoamClientOptions);

  buildGroupUrlFromId(groupId: string): string;
  hasPermission(principal: string, resourceType: string, resourceIdentifier: string, permission: string): Promise<boolean>;
  grantRoleToPrincipal(groupUrl: string, principal: string, roleName: string): Promise<void>;
  getGroupInfo(groupUrl: string): Promise<Group>;
  setAdminFlag(groupId: string, principal: string, isAdmin: boolean): Promise<void>;
  removeUserRole(groupId: string, principal: string, role: string): Promise<void>;
  addUserRole(groupId: string, principal: string, role: string): Promise<void>;
  modifyUserRoles(groupId: string, principal: string, rolesChanges: string): Promise<void>;
  addGroupMember(groupId: string, principal: string, isAdmin: boolean): Promise<void>;
  removeGroupMember(groupId: string, principal: string): Promise<void>;
  getRoles(): Promise<RoleArray>;
  findPrincipals(query: string): Promise<CanonicalPrincipalPrimitiveArrayObject>;
  getPrincipal(principal: string): Promise<Principal>;
  createGroup(name: string, description: string): Promise<Group>;
  removeGroup(groupId: string): Promise<void>;
  findGroups(resourceType: string, resourceIdentifier: string): Promise<GroupPrimitiveArray>;
  removeResourceFromGroup(groupId: string, resourceType: string, resourceId: string): Promise<ResourceArray>;
  addResourceToGroup(groupId: string, resourceType: string, resourceId: string): Promise<ResourceArray>;
  getUserPermissionsForResourceType(principal: string, resourceType: string): Promise<ResourceIdentifierPermissions[]>;
  getUsersWithPermission(resourceType: string, resourceIdentifier: string, permission: string): Promise<string[]>;
  getUsersWithResource(resourceType: string, permissionFilters: string[]): Promise<ByResourceSearchResult>;
  createGroupWithUser(principalToCreateGroup: string, principalToAddToGroup: string, groupName: string, groupDescription: string, rolesToAdd: string[], resourcesToAdd: { resourceType: string, resourceId: string }[]): Promise<string>;
}

export declare function buildGroupUrlFromId(groupId: string): string;
export declare function hasPermission(accessToken: string, principal: string, resourceType: string, resourceIdentifier: string, permission: string): Promise<boolean>;
export declare function grantRoleToPrincipal(accessToken: string, groupUrl: string, principal: string, roleName: string): Promise<void>;
export declare function getGroupInfo(accessToken: string, groupUrl: string): Promise<Group>;
export declare function setAdminFlag(accessToken: string, groupId: string, principal: string, isAdmin: boolean): Promise<void>;
export declare function removeUserRole(accessToken: string, groupId: string, principal: string, role: string): Promise<void>;
export declare function addUserRole(accessToken: string, groupId: string, principal: string, role: string): Promise<void>;
export declare function modifyUserRoles(accessToken: string, groupId: string, principal: string, rolesChanges: string): Promise<void>;
export declare function addGroupMember(accessToken: string, groupId: string, principal: string, isAdmin: boolean): Promise<void>;
export declare function removeGroupMember(accessToken: string, groupId: string, principal: string): Promise<void>;
export declare function getRoles(accessToken: string): Promise<RoleArray>;
export declare function findPrincipals(accessToken: string, query: string): Promise<CanonicalPrincipalPrimitiveArrayObject>;
export declare function getPrincipal(accessToken: string, principal: string): Promise<Principal>;
export declare function createGroup(accessToken: string, name: string, description: string): Promise<Group>;
export declare function removeGroup(accessToken: string, groupId: string): Promise<void>;
export declare function findGroups(accessToken: string, resourceType: string, resourceIdentifier: string): Promise<GroupPrimitiveArray>;
export declare function removeResourceFromGroup(accessToken: string, groupId: string, resourceType: string, resourceId: string): Promise<ResourceArray>;
export declare function addResourceToGroup(accessToken: string, groupId: string, resourceType: string, resourceId: string): Promise<ResourceArray>;
export declare function getUserPermissionsForResourceType(accessToken: string, principal: string, resourceType: string): Promise<ResourceIdentifierPermissions[]>;
export declare function getUsersWithPermission(accessToken: string, resourceType: string, resourceIdentifier: string, permission: string): Promise<string[]>;
export declare function getUsersWithResource(accessToken: string, resourceType: string, permissionFilters: string[]): Promise<ByResourceSearchResult>;
export declare function createGroupWithUser(accessToken: string, principalToCreateGroup: string, principalToAddToGroup: string, groupName: string, groupDescription: string, rolesToAdd: string[], resourcesToAdd: { resourceType: string, resourceId: string }[]): Promise<string>;


src/index.js
const CoamClient = require('./CoamClient');

const {getRoles,
    modifyUserRoles,
    removeUserRole,
    addUserRole,
    grantRoleToPrincipal,
    setAdminFlag,
    createGroup,
    removeGroup,
    findGroups,
    getGroupInfo,
    addGroupMember,
    removeGroupMember,
    group56,
    buildGroupUrlFromId,
    addResourceToGroup,
    removeResourceFromGroup,
    hasPermission,
    getUserPermissionsForResourceType,
    findPrincipals,
    getPrincipal} = require('./coamHelpers');

export {
    CoamClient,

    getRoles,

    grantRoleToPrincipal,
    addUserRole,
    removeUserRole,
    modifyUserRoles,

    setAdminFlag,

    createGroup,
    removeGroup,
    getGroupInfo,
    addGroupMember,
    removeGroupMember,

    group56,
    buildGroupUrlFromId,

    addResourceToGroup,
    removeResourceFromGroup,

    hasPermission,
    getUserPermissionsForResourceType,

    findGroups,
    findPrincipals,
    getPrincipal,
};


src/utils.js
const delay = (milliseconds, payload) => {
    return new Promise(function(resolve) {
        setTimeout(resolve.bind(null, payload), milliseconds);
    });
};

const retryOnReject = (promise, retryConditionFn, retryAttemptsLeft = 2, msBetweenRetry = 200) => {
    if ( retryAttemptsLeft === 0) {
        return Promise.reject('Gave up!');
    }
    return promise
        .then((data) => Promise.resolve(data))
        .catch((error) => {
            if (retryConditionFn(error)) {
                return delay(msBetweenRetry).then( () => retryOnReject(promise, retryConditionFn, retryAttemptsLeft-1));
            } else {
                throw error;
            }
        });
};

const retryOn403Condition = (error) => error && error.response && error.response.status === 403;

export {
    delay,
    retryOnReject,
    retryOn403Condition,
};


src/CoamClient.js
const axios = require('axios');
let axiosRetry = require('axios-retry');
const {pope} = require('pope');

const DEFAULT_BASE_URL = 'https://api.cimpress.io';

class CoamClient {
    constructor(options) {
        const opts = options || {};
        this.baseUrl = opts.baseUrl || DEFAULT_BASE_URL;
        this.timeout = Number.isInteger(opts.timeout) ? opts.timeout : 8000;
        this.accessToken = opts.accessToken || undefined;
        this.retryAttempts = Number.isInteger(opts.retryAttempts) ? opts.retryAttempts : 2;
        this.retryDelayInMs = Number.isInteger(opts.retryDelayInMs) ? opts.retryDelayInMs : 200;
        this.retryOnForbidden = 'retryOnForbidden' in opts ? opts.retryOnForbidden : true;
        this.debugFunction = 'debugFunction' in opts ? opts.debugFunction : undefined;
        // eslint-disable-next-line
        this.errorFunction = 'errorFunction' in opts ? opts.errorFunction : console.error;
        this.skipCache = 'skipCache' in opts ? opts.skipCache : false;

        this.logPrefix = '[CoamClient]';

        const validOptions = ['baseUrl', 'accessToken', 'timeout',
            'retryAttempts', 'retryDelayInMs', 'retryOnForbidden',
            'debugFunction', 'errorFunction', 'skipCache'];

        Object.keys(opts).forEach((passedOption) => {
            if (validOptions.indexOf(passedOption) === -1) {
                // eslint-disable-next-line no-console
                this.__logError(`${this.logPrefix} Option '${passedOption}' is not understood and will be ignored.`);
            }
        });

        this.instance = axios.create({
            baseURL: this.baseUrl,
            timeout: this.timeout,
        });
        this.instance.interceptors.response.use((response) => response, (error) => {
            const newError = error && error.response && {
                data: error.response.data,
                status: error.response.status,
                headers: error.response.headers,
            } || error.message && {message: error.message, code: error.code, stack: error.stack} || error;
            throw newError;
        });

        if (this.retryAttempts > 0) {
            if (typeof axiosRetry !== 'function') {
                axiosRetry = axiosRetry.default;
            }

            axiosRetry(this.instance, {
                retries: this.retryAttempts,
                retryCondition: (error) => {
                    if ( axiosRetry.isNetworkOrIdempotentRequestError(error)) {
                        return true;
                    }

                    if (this.retryOnForbidden && error && error.response && error.response.status === 403) {
                        return true;
                    }
                    // do not retry
                    return false;
                },
                retryDelay: () => {
                    return this.retryDelayInMs;
                },
                shouldResetTimeout: true,
            });
        }
    }

    __exec(data) {
        if (this.skipCache && data.method === 'GET') {
            data.params = Object.assign(data.params || {}, {skipCache: Math.random()});
        }
        return this.instance
            .request(data)
            .then((response) => {
                return response.data;
            })
            .catch((err) => {
                this.__logError(err);
                throw err;
            });
    }

    __config(data) {
        return Object.assign({}, {
            headers: {
                Authorization: `Bearer ${this.accessToken}`,
            },
        }, data);
    }

    __buildUrl(urlPattern, dataToEncode) {
        const encoded = {};
        Object.keys(dataToEncode).forEach( (k) => encoded[k] = encodeURIComponent(dataToEncode[k]));
        return pope(urlPattern, encoded);
    }

    __logDebug(...args) {
        if (this.debugFunction) {
            this.debugFunction.apply(null, args);
        }
    }

    __logError(...args) {
        if (this.errorFunction) {
            this.errorFunction.apply(null, args);
        }
    }

    buildGroupUrlFromId(groupId) {
        return `${this.baseUrl}/auth/access-management/v1/groups/${groupId}`;
    }

    hasPermission(principal, resourceType, resourceIdentifier, permission) {
        const url = this.__buildUrl(
            `/auth/access-management/v1/principals/{{principal}}/permissions/{{resourceType}}/{{resourceIdentifier}}/{{permission}}`, {
                principal,
                resourceType,
                resourceIdentifier,
                permission,
            });

        const data = this.__config({
            url: url,
            method: 'GET',
        });

        return this.__exec(data)
            .then(() => Promise.resolve(true))
            .catch((e) => {
                // COAM returns a 404 if the principal does not have access to a resource
                if (e.status === 404) {
                    return false;
                }

                // for all other errors, we want to throw an exception so that clients
                // can retry instead of assuming the user does not have access
                throw e;
            });
    }

    grantRoleToPrincipal(groupUrl, principal, roleName) {
        return this.getGroupInfo(groupUrl)
            .then((groupInfo) => {
                if (!groupInfo) {
                    return Promise.reject(new Error('Failed to retrieve template COAM group.'));
                }
                return this.addGroupMember(groupInfo.id, principal, false)
                    .then(() => this.modifyUserRoles(groupInfo.id, principal, {'add': [roleName]}));
            });
    }

    getGroupInfo(groupUrl) {
        const data = this.__config({
            url: groupUrl,
            method: 'GET',
            params: {
                canonicalize: 'true',
            },
        });

        return this.__exec(data);
    }

    setAdminFlag(groupId, principal, isAdmin) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/members/{{principal}}`, {
            groupId,
            principal,
        });

        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: {
                'is_admin': isAdmin,
            },
        });

        return this.__exec(data);
    }

    removeUserRole(groupId, principal, role) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/members/{{principal}}/roles`, {
            groupId,
            principal,
        });

        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: {
                'remove': [role],
            },
        });

        return this.__exec(data);
    }

    addUserRole(groupId, principal, role) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/members/{{principal}}/roles`, {
            groupId,
            principal,
        });

        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: {
                'add': [role],
            },
        });

        return this.__exec(data);
    }

    group56(principal) {
        const url = this.__buildUrl(`/auth/access-management/v1/principals/{{principal}}/groups`, {
            principal,
        });

        const data = this.__config({
            url: url,
            method: 'GET',
        });

        return this.__exec(data).then((data) => !!data.groups.find((a) => a.id === '56'));
    }

    modifyUserRoles(groupId, principal, rolesChanges) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/members/{{principal}}/roles`, {
            groupId,
            principal,
        });

        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: rolesChanges,
        });

        return this.__exec(data);
    }

    addGroupMember(groupId, principal, isAdmin) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/members`, {
            groupId,
            principal,
        });
        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: {
                'add': [{
                    is_admin: !!isAdmin,
                    principal: principal,
                }],
            },
            params: {
                canonicalize: true,
            },
        });

        return this.__exec(data);
    }

    removeGroupMember(groupId, principal) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/members`, {
            groupId,
        });
        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: {
                'remove': [principal],
            },
        });

        return this.__exec(data);
    }

    getRoles() {
        const data = this.__config({
            url: `/auth/access-management/v1/roles`,
            method: 'GET',
        });

        return this.__exec(data);
    }

    findPrincipals(query) {
        if (!query || query.length === 0) {
            return Promise.resolve([]);
        }

        const data = this.__config({
            url: '/auth/access-management/v1/search/canonicalPrincipals/bySubstring',
            method: 'GET',
            params: {
                q: query,
                canonicalize: true,
            },
        });

        // [{user_id / name / email}]
        return this.__exec(data).then((p) => p.canonical_principals);
    }

    getPrincipal(principal) {
        const url = this.__buildUrl(`/auth/access-management/v1/principals/{{principal}}`, {
            principal,
        });
        const data = this.__config({
            url,
            method: 'GET',
        });
        return this.__exec(data);
    }

    createGroup(name, description) {
        const data = this.__config({
            url: '/auth/access-management/v1/groups',
            method: 'POST',
            data: {
                name: name,
                description: description,
            },
        });

        // [{user_id / name / email}]
        return this.__exec(data);
    }

    removeGroup(groupId) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}`, {
            groupId,
        });

        const data = this.__config({
            url: url,
            method: 'DELETE',
        });

        return this.__exec(data);
    }

    findGroups(resourceType, resourceIdentifier) {
        const data = this.__config({
            url: `/auth/access-management/v1/groups`,
            method: 'GET',
            params: {
                resource_type: resourceType,
                resource_identifier: resourceIdentifier,
            },
        });

        return this.__exec(data);
    }

    removeResourceFromGroup(groupId, resourceType, resourceId) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/resources`, {
            groupId,
        });

        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: {
                add: [],
                remove: [{
                    resource_type: resourceType,
                    resource_identifier: resourceId,
                }],
            },
        });

        return this.__exec(data);
    }

    addResourceToGroup(groupId, resourceType, resourceId) {
        const url = this.__buildUrl(`/auth/access-management/v1/groups/{{groupId}}/resources`, {
            groupId,
        });

        const data = this.__config({
            url: url,
            method: 'PATCH',
            data: {
                add: [{
                    resource_type: resourceType,
                    resource_identifier: resourceId,
                }],
                remove: [],
            },
        });

        return this.__exec(data);
    }

    getUserPermissionsForResourceType(principal, resourceType, include = true, permissionFilters = null) {
        const urlString = `/auth/access-management/v1/principals/{{principal}}/permissions/{{resourceType}}?include=${include}`.concat(permissionFilters ? `&permissionFilter=${permissionFilters.join(',')}`:'');
        const url = this.__buildUrl(urlString, {
            principal,
            resourceType,
        });

        const data = this.__config({
            url: url,
            method: 'GET',
        });

        return this.__exec(data);
    }

    getUserPermissionsForResourceTypes(principal, resourceTypes, include = true, permissionFilters = null) {
        const urlString = `/auth/access-management/v1/principals/{{principal}}/permissions?resourceTypes=${resourceTypes.join(',')}&include=${include}`.concat(permissionFilters ? `&permissionFilter=${permissionFilters.join(',')}`:'');
        const url = this.__buildUrl(urlString, {
            principal,
        });

        const data = this.__config({
            url: url,
            method: 'GET',
        });

        return this.__exec(data);
    }

    getUsersWithPermission(resourceType, resourceIdentifier, permission) {
        const url = this.__buildUrl(`/auth/access-management/v1/search/canonicalPrincipals/byPermission?resource_type={{resourceType}}${resourceIdentifier ? '&resource_identifier={{resourceIdentifier}}' : ''}&permission={{permission}}`, {
            resourceType,
            resourceIdentifier,
            permission,
        });

        const data = this.__config({
            url: url,
            method: 'GET',
        });

        // [{user_id / name / email}]
        return this.__exec(data).then((p) => p.canonical_principals);
    }


    getUsersWithResource(resourceType, permissionFilters) {
        const url = this.__buildUrl(`/auth/access-management/v1/search/canonicalPrincipals/byResource?resource_type={{resourceType}}${permissionFilters && permissionFilters.length > 0 ? '&permissionFilter={{commaSeparatedPermissionFilter}}' : ''}`, {
            resourceType,
            commaSeparatedPermissionFilter: permissionFilters && permissionFilters.length > 0 && permissionFilters.join(','),
        });

        const data = this.__config({
            url: url,
            method: 'GET',
        });

        // [{canonical_principal / resource_type / is_client / permissions}]
        return this.__exec(data);
    }

    async createGroupWithUser(principalToCreateGroup, principalToAddToGroup, groupName, groupDescription, rolesToAdd, resourcesToAdd) {
        const res = await this.createGroup(groupName, groupDescription);
        const groupId = res.id;

        await this.addGroupMember(groupId, principalToAddToGroup);
        await this.setAdminFlag(groupId, principalToAddToGroup, true);
        for (let i = 0; i < rolesToAdd.length; i++) {
            await this.addUserRole(groupId, principalToAddToGroup, rolesToAdd[i]);
        }
        for (let i = 0; i < resourcesToAdd.length; i++) {
            const resource = resourcesToAdd[i];
            await this.addResourceToGroup(groupId, resource.resourceType, resource.resourceId);
        }
        try {
            await this.removeGroupMember(groupId, principalToCreateGroup);
        } catch (err) {
            //  Removing the creator of a COAM group causes a 404, known issue
            if (err.status !== 404) {
                throw err;
            }
        }
        return groupId;
    }
}

module.exports = CoamClient;
